bindings/azure/eventgrid/eventgrid.go

```diff
@@ -0,0 +1,191 @@
+	SubscriptionID            string `json:"subscriptionId"`
+	ResourceGroupName         string `json:"resourceGroupName"`
+	TopicName                 string `json:"topicName"`
+	EventGridSubscriptionName string `json:"eventGridSubscriptionName"`
+	SubscriberEndpoint        string `json:"subscriberEndpoint"`
+	TopicEndpoint             string `json:"topicEndpoint"`
+	ClientID                  string `json:"clientId"`
+	ClientSecret              string `json:"clientSecret"`
+	HandshakePort             string `json:"handshakePort"`
+}
+
+// NewAzureEventGrid returns a new Azure Event Grid instance
+func NewAzureEventGrid(logger logger.Logger) *AzureEventGrid {
+	logger.Debug("NewAzureEventGrid() called...")
+	return &AzureEventGrid{logger: logger}
+}
+
+// Init performs metadata init
+func (a *AzureEventGrid) Init(metadata bindings.Metadata) error {
+	a.logger.Debugf("Parsing Event Grid metadata(%s)...", metadata.Name)
+
+	m, err := a.parseMetadata(metadata)
+	if err != nil {
+		return err
+	}
+	a.metadata = m
+
+	a.logger.Debug("Metadata parsed successfully.")
+
+	return nil
+}
+
+func (a *AzureEventGrid) Read(handler func(*bindings.ReadResponse) error) error {
+	a.logger.Debug("Read() called...")
+
+	err := a.createSubscription()
+	if err != nil {
+		return err
+	}
+
+	m := func(ctx *fasthttp.RequestCtx) {
+		switch string(ctx.Path()) {
+		case "/api/events":
+			if string(ctx.Method()) == "OPTIONS" {
+				ctx.Response.Header.Add("WebHook-Allowed-Origin", string(ctx.Request.Header.Peek("WebHook-Request-Origin")))
+				ctx.Response.Header.Add("WebHook-Allowed-Rate", "*")
+				ctx.Response.Header.SetStatusCode(fasthttp.StatusOK)
+				_, err = ctx.Response.BodyWriter().Write([]byte(""))
+				if err != nil {
+					a.logger.Error(err.Error())
+				}
+			} else if string(ctx.Method()) == "POST" {
+				bodyBytes := ctx.PostBody()
+
+				a.logger.Debug(string(bodyBytes))
+				err = handler(&bindings.ReadResponse{
+					Data: bodyBytes,
+				})
+				if err != nil {
+					a.logger.Error(err.Error())
+				}
+			}
+		}
+	}
+
+	fasthttp.ListenAndServe(fmt.Sprintf(":%s", a.metadata.HandshakePort), m)
+
+	a.logger.Debugf("listening for Event Grid events at http://localhost:%s/api/events", a.metadata.HandshakePort)
+
+	return nil
+}
+
+func (a *AzureEventGrid) Write(req *bindings.WriteRequest) error {
+	a.logger.Debug("Write() called. data: %s", string(req.Data))
+
+	request := fasthttp.AcquireRequest()
+	defer fasthttp.ReleaseRequest(request)
+	request.Header.SetMethod(fasthttp.MethodPost)
+	request.Header.Set("Content-Type", "application/cloudevents+json")
+	request.Header.Set("aeg-sas-key", a.metadata.ClientSecret)
+	request.SetRequestURI(a.metadata.TopicEndpoint)
+	request.SetBody(req.Data)
+
+	response := fasthttp.AcquireResponse()
+	defer fasthttp.ReleaseResponse(response)
+
+	client := &fasthttp.Client{WriteTimeout: time.Second * 10}
+	err := client.Do(request, response)
+	if err != nil {
+		a.logger.Error(err.Error())
```

Comment:
Shouldn't this be handled just like the case for statuscode != OK?

Reply:
Now properly returning the error.